networks/2_module_validation.py

```diff
@@ -0,0 +1,300 @@
+    """
+    plots cumulative (or survival) step distribution
+    adapted from https://github.com/MarvinT/morphs/blob/master/morphs/plot/utils.py
+    """
+    data = np.sort(data)
+    if survival:
+        data = data[::-1]
+    y = np.arange(data.size + 1, dtype=float)
+    if scaled:
+        y /= y[-1]
+    x = np.concatenate([data, data[[-1]]])
+    plt.step(x, y, label=label, **kwargs)
+    if fill:
+        plt.fill_between(x, y, alpha=0.5, step="pre", **kwargs)
+
+
+# +
+cumulative_distribution(
+    pao1_operon["Num_operon_modules"],
+    label="Number of modules containing operon genes",
+    color="red",
+)
+cumulative_distribution(
+    pao1_operon["Num_random_modules"],
+    label="Number of modules containing random genes",
+    color="blue",
+)
+_ = plt.legend()
+plt.title("Distribution of module counts (operon vs random genes)")
+
+scipy.stats.ks_2samp(
+    pao1_operon["Num_operon_modules"], pao1_operon["Num_random_modules"]
+)
+# -
+
+fig = sns.displot(
+    pao1_operon_toplot,
+    x="value",
+    hue="variable",
+)
+plt.title("Distribution of module counts (operon vs random genes)")
+
+# +
+cumulative_distribution(
+    pao1_regulon["Num_regulon_modules"],
+    label="Number of modules containing regulon genes",
+    color="red",
+)
+cumulative_distribution(
+    pao1_regulon["Num_random_modules"],
+    label="Number of modules containing random genes",
+    color="blue",
+)
+_ = plt.legend()
+plt.title("Distribution of module counts (regulon vs random genes)")
+
+scipy.stats.ks_2samp(
+    pao1_regulon["Num_regulon_modules"], pao1_regulon["Num_random_modules"]
+)
+# -
+
+fig = sns.displot(
+    pao1_regulon_toplot,
+    x="value",
+    hue="variable",
+)
+plt.title("Distribution of module counts (regulon vs random genes)")
+
+# **Takeaway:**
+# * The axis cumulative distribution plots are:
+#     * x: number of modules that operons/regulons/random genes are contained in
+#     * y: count
+# * These distribution plots are summing counts as you move from left to right, so a shift in the curves corresponds to a shift in the distribution (i.e. a curve shifted to the right means that the distribution is shifted to the right)
+#
+# * We can perform [Kolmogorov-Smirnov test](https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test) to compare the distribution of module counts for genes in regulons/operons versus random genes. The KS test will quantify the difference in the cumulative distribution curves.
+# * Based on the KS test, there is a significant difference between the two distributions as we would expect. Though there is a very large significance for the operons due to the large sample size. There are~3K operons as opposed to the 17 regulons, which has a very modestly significant p-value. This difference is fairly consistent across thresholds, though as we decrease the threshold used to create the network, the significance for the operons increases slightly but the significance for the regulons decreases slightly.
```

Comment:
This is what I would expect too, but I wonder how much of this is driven by the single large module. It might be interesting to look at the number of single-module operons and regulons that are in the largest module, as opposed to those where the genes are all contained in a smaller module.

Subjectively I'd find the relationships to be more meaningful if the module is smaller (e.g. if an operon with 5 genes is contained in a module consisting of 10 total genes, this seems more biologically/functionally meaningful than an operon with 5 genes contained in a module consisting of 500 genes).

I guess maybe you could make this quantitative with something like a hypergeometric test, but a simple/straightforward first step might just be looking at the sizes of the modules that contain whole operons or regulons.

Reply:
This is a very good point. I will look at the size of the modules. Another thing I can try is to just remove the modules with >1K genes and see what the trends look like too.

Reply:
AFTER removing modules with > 1000 genes
* Using the KS test, there is a significant difference (across thresholds) between the operon and random distribution, as we would expect.
* There is only a significant difference between the regulon and random distributions at lower thresholds (0.7 and below). The lack of significance at higher thresholds is likely due to the small sample size (i.e. 6 or 10 regulons have genes that are contained in modules with fewer than 1000 genes)